### PR TITLE
Support masking of unconnected nodes (LAM)

### PR DESCRIPTION
**Pull request by JPXKQX**
This PR adds support for using masked graphs, resulting in a performance improvement in LAM by dropping unused nodes in the boundary forcing dataset.

## Description
The current release of anemoi-graphs (`0.4.1`) supports dropping unconnected nodes from the limited area graph, with the following code:
```
post_processors:
  - _target_: anemoi.graphs.processors.RemoveUnconnectedNodes
    nodes_name: data
    ignore: cutout_mask # optional
    save_mask_indices_to_attr: indices_connected_nodes # optional
```
`anemoi-graphs` returns the masked graph with an additional node attribute `indices_connected_nodes` (in this example) representing the indices used to mask the input dataset to the new graph `data` nodes.

## Solution

We need to specify these spatial indices in the dataloader config. 
```
grid_indices:
    _target_: anemoi.training.data.grid_indices.MaskedGrid
    nodes_name: data
    node_attribute_name: indices_connected_nodes
```

The performance improvement is of 66 % (from 1,92 it/s to 3,19 it/s) for the 1 GPU default limited area model.

Screenshot 2024-11-28 at 10 58 06

This PR has been tested successfully on 1 GPU (with `plot: detailed` without `PlotSpectrum` included): 
- [x] Limited Area Model with `reader_group_size=1`
- [x] Stretched grid with `reader_group_size=1`
- [x] Global model with `reader_group_size=1`

This PR has been tested successfully on 2 GPU (with `plot: detailed` without `PlotSpectrum` included): 
- [x] Limited Area Model with `reader_group_size=2`
- [x] Stretched grid with `reader_group_size=2`
- [x] Global model with `reader_group_size=2`

W.I.P.: This mask must be saved to the checkpoint so that it can be used in `anemoi-inference`.

### CODE REVIEW

**Review by b8raoult — DISMISSED** [review]

File: src/anemoi/training/data/dataset.py [hunk 1]
```diff
@@ -157,14 +163,10 @@ def set_comm_group_info(
...
-            grid_shard_size = self.grid_size // self.reader_group_size
-            self.grid_start = self.reader_group_rank * grid_shard_size
-            if self.reader_group_rank == self.reader_group_size - 1:
-                self.grid_end = self.grid_size
-            else:
-                self.grid_end = (self.reader_group_rank + 1) * grid_shard_size
+        assert self.reader_group_size >= 1, "reader_group_size must be postive"
+
+        # get the grid shard size and start/end indices
+        self.grid_shard_indices = self.grid_indices.get_shard_indices(self.reader_group_rank, self.reader_group_size)
```
[the remark this answers is not in the document]
Comment (JPXKQX) [reply src/anemoi/training/data/dataset.py]: True

File: src/anemoi/training/data/dataset.py [hunk 2]
```diff
@@ -157,14 +163,7 @@ def set_comm_group_info(
...
 
-        if self.reader_group_size > 1:
-            # get the grid shard size and start/end indices
-            grid_shard_size = self.grid_size // self.reader_group_size
-            self.grid_start = self.reader_group_rank * grid_shard_size
-            if self.reader_group_rank == self.reader_group_size - 1:
-                self.grid_end = self.grid_size
-            else:
-                self.grid_end = (self.reader_group_rank + 1) * grid_shard_size
+        assert self.reader_group_size >= 1, "reader_group_size must be postive"
```
Comment (theissenhelen) [inline src/anemoi/training/data/dataset.py:166]: change to "positive"

**Review by theissenhelen — CHANGES_REQUESTED** [review]
Just a typo to fix.

**Review by theissenhelen — APPROVED** [review]
Great work!

**Pull request merged**